### PR TITLE
[onert] Introduce RUY backend

### PR DESCRIPTION
**Pull request by periannath**
- This commit introduces RUY backend
  - Now it only supports FP32 Conv2D operation
  - Most contents are copied from CPU backend

ONE-DCO-1.0-Signed-off-by: JiHwan Yeo <jihwan.yeo@samsung.com>

---

From draft #4870 
For issue #4863

### CODE REVIEW

File: runtime/onert/backend/ruy/ExternalContext.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+#ifndef __ONERT_BACKEND_RUY_EXTERNAL_CONTEXT_H__
+#define __ONERT_BACKEND_RUY_EXTERNAL_CONTEXT_H__
+
+#include <util/ConfigSource.h>
+#include <ruy/context.h>
+
+namespace
+{
+const int kDefaultNumThreadpoolThreads = 4;
```
Comment (periannath) [inline runtime/onert/backend/ruy/ExternalContext.h:25]: I've set default number of threads for ruy as 4. This value shows best performance. However, cpu backend uses 1 thread. I'm not sure it is okay to use 4 threads as default value.

**Comment by periannath** [comment]
@nnfw-bot test cross-runtime-test-debug

**Comment by periannath** [comment]
@nnfw-bot test cross-runtime-test-release

**Comment by periannath** [comment]
## Test result
- Model : MobileNetV2

### RUY backend
```bash
$ OP_BACKEND_Conv2D=ruy BACKENDS="cpu;ruy" ./Product/armv7l-linux.release/out/bin/nnpackage_run --nnpackage ../nnpkg_tst/mobilenet_v2_1.0_224 -w1 -r20
Package Filename ../nnpkg_tst/mobilenet_v2_1.0_224
===================================
MODEL_LOAD   takes 43.531 ms
PREPARE      takes 23.696 ms
EXECUTE      takes 72.894 ms
- MEAN     :  72.894 ms
- MAX      :  87.957 ms
- MIN      :  71.400 ms
- GEOMEAN  :  72.812 ms
===================================
```

### CPU backend
```bash
$ OP_BACKEND_Conv2D=cpu BACKENDS="cpu;ruy" ./Product/armv7l-linux.release/out/bin/nnpackage_run --nnpackage ../nnpkg_tst/mobilenet_v2_1.0_224 -w1 -r20
Package Filename ../nnpkg_tst/mobilenet_v2_1.0_224
===================================
MODEL_LOAD   takes 41.401 ms
PREPARE      takes 84.211 ms
EXECUTE      takes 134.922 ms
- MEAN     :  134.922 ms
- MAX      :  138.324 ms
- MIN      :  132.527 ms
- GEOMEAN  :  134.912 ms
===================================
```

File: runtime/onert/backend/ruy/ops/ConvolutionLayer.cc [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  nnfw::ruy::Conv &kernel = *_conv_kernel;
+  if (_input->data_type() == OperandType::FLOAT32 && _kernel->is_constant())
+  {
+    kernel.prepare(getTensorShape(_input), getTensorShape(_kernel), getTensorShape(_output),
+                   _strideWidth, _strideHeight, _dilationWidthFactor, _dilationHeightFactor);
+  }
+  _prepare = true;
+}
+
+#undef ANDROID_NN_CONV_PARAMETERS
```
Comment (YongseopKim) [inline runtime/onert/backend/ruy/ops/ConvolutionLayer.cc:150]: ?

File: runtime/onert/backend/ruy/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,22 @@
+set(LIB_ONERT_BACKEND_RUY onert_backend_ruy)
+
+nnfw_find_package(Ruy REQUIRED)
+
+file(GLOB_RECURSE SOURCES "*.cc")
+
+add_library(${LIB_ONERT_BACKEND_RUY} SHARED ${SOURCES})
+
+target_link_libraries(${LIB_ONERT_BACKEND_RUY} PRIVATE nnfw_lib_ruy)
```
Comment (YongseopKim) [inline runtime/onert/backend/ruy/CMakeLists.txt:9]: What difference btw this and below ruy?

Comment (periannath) [reply runtime/onert/backend/ruy/CMakeLists.txt:9]: @YongseopKim `nnfw_lib_ruy` means kernel implementations that use ruy library. It is located in `compute/ruy`. `ruy` means external `RUY` library itself.

File: runtime/onert/backend/ruy/ops/ConvolutionLayer.cc (hunk 2 above)
Comment (periannath) [reply runtime/onert/backend/ruy/ops/ConvolutionLayer.cc:150]: @YongseopKim Unnecessary legacy code is copied from CPU backend. I'll remove it.

**Comment by glistening** [comment]
> ## Test result
> 
>     * Model : MobileNetV2
>     * Device : Odroid-XU4
> 
> 
> ### RUY backend
> 
> ```shell
> $ OP_BACKEND_Conv2D=ruy BACKENDS="cpu;ruy" ./Product/armv7l-linux.release/out/bin/nnpackage_run --nnpackage ../nnpkg_tst/mobilenet_v2_1.0_224 -w1 -r20
> Package Filename ../nnpkg_tst/mobilenet_v2_1.0_224
> ===================================
> MODEL_LOAD   takes 43.531 ms
> PREPARE      takes 23.696 ms
> EXECUTE      takes 72.894 ms
> - MEAN     :  72.894 ms
> - MAX      :  87.957 ms
> - MIN      :  71.400 ms
> - GEOMEAN  :  72.812 ms
> ===================================
> ```
> 
> ### CPU backend
> 
> ```shell
> $ OP_BACKEND_Conv2D=cpu BACKENDS="cpu;ruy" ./Product/armv7l-linux.release/out/bin/nnpackage_run --nnpackage ../nnpkg_tst/mobilenet_v2_1.0_224 -w1 -r20
> Package Filename ../nnpkg_tst/mobilenet_v2_1.0_224
> ===================================
> MODEL_LOAD   takes 41.401 ms
> PREPARE      takes 84.211 ms
> EXECUTE      takes 134.922 ms
> - MEAN     :  134.922 ms
> - MAX      :  138.324 ms
> - MIN      :  132.527 ms
> - GEOMEAN  :  134.912 ms
> ===================================
> ```

@periannath According to https://github.com/Samsung/ONE/pull/5067#discussion_r524920361, `CPU backend` uses **1** thread, whileas `Ruy backend` uses **4**  threads. Right?

**Comment by periannath** [comment]
@glistening Here is the result with 1 ruy thread.

```bash
$ RUY_THREADS=1 OP_BACKEND_Conv2D=ruy BACKENDS="cpu;ruy" ./Product/armv7l-linux.release/out/bin/nnpackage_run --nnpackage ../nnpkg_tst/mobilenet_v2_1.0_224 -w1 -r20
Package Filename ../nnpkg_tst/mobilenet_v2_1.0_224
===================================
MODEL_LOAD   takes 41.801 ms
PREPARE      takes 23.743 ms
EXECUTE      takes 134.833 ms
- MEAN     :  134.833 ms
- MAX      :  141.096 ms
- MIN      :  134.038 ms
- GEOMEAN  :  134.826 ms
===================================
```

- Please keep in mind that cpu backend uses 8 threads on odroid-xu4.

**Comment by seanshpark** [comment]
> Please keep in mind that Conv2D kernel on cpu backend uses 8 threads on odroid-xu4.

so, RUY 1 thread == CPU 8 thread ?

**Comment by periannath** [comment]
@seanshpark Yes. That's right on MobileNetV2.

**Comment by lemmaa** [comment]
@periannath , can we see the result of using `acl_neon` instead of cpu except for conv_2d? That is, `OP_BACKEND_Conv2D=ruy BACKENDS="acl_neon;ruy" ...`. Also in this case, the result will be different depending on the number of threads of acl_neon or the number of threads of ruy kernel.

**Comment by periannath** [comment]
@lemmaa Here are some results with `acl_neon` backend. 

|  | inference time |
|--|--|
| `acl_neon` + `ruy` for conv2d (4 threads) | 101.6 ms |
| `acl_neon` + `ruy` for conv2d (1 threads) | 125.7 ms |
| `acl_neon` | 43.8 ms |

- `acl_neon` + `ruy` has permutations between backends
  - Some permutation layer takes a few milliseconds

<details>
<summary>Raw data</summary>

1. `acl_neon` + `ruy` for conv2d (4 threads)

```
$ OP_BACKEND_Conv2D=ruy BACKENDS="acl_neon;cpu;ruy" ./Product/armv7l-linux.release/out/bin/nnpackage_run --nnpackage ../nnpkg_tst/mobilenet_v2_1.0_224 -w1 -r200
Package Filename ../nnpkg_tst/mobilenet_v2_1.0_224
===================================
MODEL_LOAD   takes 41.292 ms
PREPARE      takes 95.675 ms
EXECUTE      takes 101.629 ms
- MEAN     :  101.629 ms
- MAX      :  131.467 ms
- MIN      :  91.837 ms
- GEOMEAN  :  101.455 ms
===================================
```

2. `acl_neon` + `ruy` for conv2d (1 threads)

```bash
$ RUY_THREADS=1 OP_BACKEND_Conv2D=ruy BACKENDS="acl_neon;cpu;ruy" ./Product/armv7l-linux.release/out/bin/nnpackage_run --nnpackage ../nnpkg_tst/mobilenet_v2_1.0_224 -w1 -r200
Package Filename ../nnpkg_tst/mobilenet_v2_1.0_224
===================================
MODEL_LOAD   takes 41.407 ms
PREPARE      takes 94.441 ms
EXECUTE      takes 125.735 ms
- MEAN     :  125.735 ms
- MAX      :  134.712 ms
- MIN      :  124.448 ms
- GEOMEAN  :  125.731 ms
===================================
```

3. `acl_neon`

```bash
$ BACKENDS="acl_neon;cpu;ruy" ./Product/armv7l-linux.release/out/bin/nnpackage_run --nnpackage ../nnpkg_tst/mobilenet_v2_1.0_224 -w1 -r200
Package Filename ../nnpkg_tst/mobilenet_v2_1.0_224
===================================
MODEL_LOAD   takes 44.306 ms
PREPARE      takes 395.172 ms
EXECUTE      takes 43.824 ms
- MEAN     :  43.824 ms
- MAX      :  49.413 ms
- MIN      :  42.192 ms
- GEOMEAN  :  43.814 ms
===================================
```

</details>

**Comment by periannath** [comment]
@chunseoklee @YongseopKim PTAL

**Review by chunseoklee — APPROVED** [review]

**Review by YongseopKim — APPROVED** [review]
LGTM

**Pull request merged**